datafusion/Cargo.toml

```diff
@@ -56,9 +56,9 @@ paste = "^1.0"
 num_cpus = "1.13.0"
 chrono = "0.4"
 async-trait = "0.1.41"
-futures = "0.3"
+futures = { version = "0.3", features = ["executor"] }
```

Comment:
While using `tokio::runtime::Handle::block_on`, I'm facing with:
> ’Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.

Since `block_on` is `try_enter`ing an already entered runtime, therefore I changed to `future::executor`'s to avoid panic in the first place. But as I noted before, `future::executor::block_on` is also flawed here:

> However, this approach is flawed for block_on may block the only thread in tokio, and the future inside won't get a chance to run, therefore hanging forever if the tokio runtime is not a multi-threaded one. (I temporarily change the related test to use #[tokio::test(flavor = "multi_thread", worker_threads = 2)] to avoid hanging). 

Do you have any suggestions on this?

Reply:
I think the only real suggestion is to plumb `async` all the way through to planning (aka remove the non async API)

Reply:
> The alternate compromise, which you have partly implemented in this PR, is to implement both async and non async versions. This is similar to the approach in the C/C++ filesystem api (props to @nealrichardson for the pointer), which has both having synchronous and asynchronous APIs.

How about this alternative to reduce the scope of this PR? i.e. implement both sync and async, but only use sync API to migrate existing code to the new IO abstraction, then work on async propagation as a fast follow up.

Reply:
The other thing I was thinking about was what about adding in the ObjectStore interfaces in one PR and then start hooking that up into the rest of the system / rewrite the existing data sources (like Parquet, etc) as separate PRs. 

I think @yjshen  has done a great job with this PR showing how everything would hook together, but I do feel like this PR is slightly beyond my ability to comprehend given its size and scope.

Reply:
I am onboard with further reducing the scope by focusing only on the ObjectStore interface :)